include/mlir-tcp/Conversion/TorchToTcp/TcpCustomOpBuilder.h

```diff
@@ -0,0 +1,71 @@
+//===------------------------------------------------------------*- C++ -*-===//
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "mlir-tcp/Dialect/IR/TcpOps.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+
+namespace mlir::tcp {
+
+/// Helper class for converting torch ops to tcp.custom_ops. Encapsulates a
+/// few common patterns which we see being used in these conversions:
+///
+/// 1. Some of the tensor operands to the original torch op are converted
+/// into positional tensor operands. Given that a tcp.custom_op does not
+/// have a strongly typed API to fetch the operands other than by index, we
+/// also add an extra StringArray attribute to the tcp.custom_op called
+/// "torch_operand_names" which describes what each positional argument of
+/// the `tcp.custom_op` corresponds to. This also allows us to handle the
+/// case of optional tensor arguments.
+///
+/// 2. Some of the tensor operands are converted into named attributes of
+/// the final tcp.custom_op
+///
+/// See `ConvertAtenConvolutionOp` and `ConvertAten_IndexPutImplOp` for an
+/// example usages.
+class TcpCustomOpBuilder {
+
+public:
+  TcpCustomOpBuilder(Operation *a_op, ConversionPatternRewriter &a_rewriter,
+                     const TypeConverter *a_typeConverter)
+      : op(a_op), rewriter(a_rewriter), typeConverter(a_typeConverter) {}
+
+  /// Add a value as a named tensor operand.
+  void addOperand(std::string opName, Value value);
+
+  /// Expand the passed in value as multiple named tensor operands. Expects
+  /// that value is a torch list type with tensors as elements
+  void addAsMultipleTensorOperands(std::string opNamePrefix, mlir::Value value);
+
+  // Add value as a named bool attribute
+  void addBoolAttr(std::string attrName, Value value);
+
+  // Add value as a named integer attribute
+  void addIntAttr(std::string attrName, Value value);
+
+  // Add value as a named list of integers attribute
+  void addListOfIntsAttr(std::string attrName, Value value);
```

Comment:
Do you have any strong reasons to have these as methods on a class as compared to stand-alone functions?

Here are the advantages I see with stand-alone functions:
* These functions can be used through this conversion code (not just the custom op conversions) without instantiating any objects. We don't have to keep track of the object lifetime in that case.
* Builder is already over-loaded in MLIR. Adding one more here seems confusing.
* That is the approach followed in MLIR AFAICT. So, there will be uniformity.

So, I strongly recommend moving to stand-alone functions.

Reply:
@sjain-stanford Can you think of other reasons for going with stand-alone functions?

Reply:
The reason I went with methods is that as we call them, we are accumulating the state of the built-up tcp.custom_op's operands etc. into `operandNames`, `operands`, `conversionResult`, `attrs`. Without these things being methods on some class, all of that state will have to be hoisted into the callee function and passed in. At that point, the code basically becomes "classes in C" as opposed to a C++ class. At least I cannot immediately imagine writing the callee-side code as cleanly without this encapsulation. See for instance, this bit:

```cpp
    builder.addOperand("self", adaptor.getSelf());
    builder.addOperand("index", adaptor.getIndex());
    builder.addIntAttr("axis", op.getDim());

    return builder.replace();
```

Without the class, would we have to declare the members of the class in that function (operandNames etc.) and then pass these in by reference to these (now free) functions? How would that be really different?